File: data-structures/hash-table/typescript/hash-table.ts
export class HashTable<T> {
	private _keyMap: Array<[string, T][]>;

	constructor(size: number) {
		this._keyMap = new Array(size);
	}

	set(key: string, value: T): void {
		const index = this._hash(key);

		if (!this._keyMap[index]) {
			this._keyMap[index] = [];
		}

		this._keyMap[index].push([key, value]);
	}

	get(key: string): T | undefined {
		const index = this._hash(key);

		const bucket = this._keyMap[index];

		if (bucket) {
			for (let pair of bucket) {
				if (pair[0] === key) {
					return pair[1];
				}
			}
		}

		return undefined;
	}

	remove(key: string): void {
		const index = this._hash(key);
		if (this._keyMap[index]) {
			this._keyMap[index] = this._keyMap[index].filter(pair => {
				pair[0] === key;
			});
		}
	}

	keys(): string[] {
		const keys: string[] = [];
		for (let bucket of this._keyMap) {
			if (bucket) {
				bucket.forEach(pair => {
					keys.push(pair[0]);
				});
			}
		}

		return keys;
	}

	values(): T[] {
		const values: T[] = [];
		for (let bucket of this._keyMap) {
			if (bucket) {
				bucket.forEach(pair => {
					values.push(pair[1]);
				});
			}
		}

		return values;
	}

	private _hash(key: string) {
		let total = 0;
		let PRIME = 31; // A prime number to help with distribution
		for (let i = 0; i < Math.min(key.length, 100); i++) {
			let char = key[i];
			let value = char.charCodeAt(0) - 96;
			total = (total * PRIME + value) % this._keyMap.length;
		}
		return total;
	}
}

File: data-structures/hash-table/python/hash_table.py
class HashTable:
    def __init__(self, size):
        self._key_map = [None] * size

    def set(self, key, value):
        index = self._hash(key)
        if not self._key_map[index]:
            self._key_map[index] = []
        self._key_map[index].append((key, value))

    def get(self, key):
        index = self._hash(key)
        bucket = self._key_map[index]
        if bucket:
            for pair in bucket:
                if pair[0] == key:
                    return pair[1]
        return None

    def remove(self, key):
        index = self._hash(key)
        if self._key_map[index]:
            self._key_map[index] = [pair for pair in self._key_map[index] if pair[0] != key]

    def keys(self):
        keys = []
        for bucket in self._key_map:
            if bucket:
                for pair in bucket:
                    keys.append(pair[0])
        return keys

    def values(self):
        values = []
        for bucket in self._key_map:
            if bucket:
                for pair in bucket:
                    values.append(pair[1])
        return values

    def _hash(self, key):
        total = 0
        prime = 31
        for i in range(min(len(key), 100)):
            char = key[i]
            value = ord(char) - 96
            total = (total * prime + value) % len(self._key_map)
        return total
